README.md
<span align="center">

<a href="https://github.com/homebridge/verified/blob/master/verified-plugins.json"><img alt="homebridge-verified" src="https://raw.githubusercontent.com/homebridge-plugins/homebridge-honeywell-home/master/honeywell/Homebridge_x_Honeywell.svg?sanitize=true" width="500px"></a>

# Homebridge Honeywell Home Room Sensor

<a href="https://www.npmjs.com/package/homebridge-honeywell-home-roomsensor"><img title="npm version" src="https://badgen.net/npm/v/homebridge-honeywell-home-roomsensor" ></a>
<a href="https://www.npmjs.com/package/homebridge-honeywell-home-roomsensor"><img title="npm downloads" src="https://badgen.net/npm/dt/homebridge-honeywell-home-roomsensor" ></a>

The **Homebridge Honeywell Home Room Sensor** plugin is and add-on to the [**Homebridge Honeywell Home**](https://www.npmjs.com/package/homebridge-honeywell-home) plugin.

**Homebridge Honeywell Home Room Sensor** plugin allows you to access your T9 Honeywell Home thermostat room sensor(s) from HomeKit.

</p>

</span>

## Installation

1. Search for "Honeywell Home Roomsensor" on the plugin screen of [Homebridge Config UI X](https://github.com/oznu/homebridge-config-ui-x).
2. Click **Install**.

## Configuration

1. Login / create an account at https://developer.honeywellhome.com/user
2. Click **Create New App**
3. Give your application a name, and enter the Callback URL as `https://homebridge-honeywell.iot.oz.nu/link-account`
4. Enter the generated consumer key and secret into the plugin settings screen of [Homebridge Config UI X](https://github.com/oznu/homebridge-config-ui-x)
5. Click **Link Account**

  <p align="center">

  <img src="https://user-images.githubusercontent.com/9875439/90537529-0b7cad00-e143-11ea-8c49-6820ccdec3bb.png" width="600px">

  </p>

## Homebridge Honeywell Home Plugins

- [homebridge-honeywell-home](https://github.com/donavanbecker/homebridge-honeywell-home)
  - The Homebridge Honeywell Home plugin allows you to access your Honeywell Home thermostat from HomeKit.
  - This Plugin is the main plugin that all other plugins root from. This combines all the plugins together.
- [homebridge-honeywell-home-thermostat](https://github.com/donavanbecker/homebridge-honeywell-home-thermostat)
  - The Homebridge Honeywell Home Thermsotat plugin allows you to control your Honeywell Home thermostats from HomeKit.
- [homebridge-honeywell-home-roomsensor](https://github.com/donavanbecker/homebridge-honeywell-home-roomsensor)
  - The Homebridge Honeywell Home Room Sensor plugin allows you to access your T9 Honeywell Home thermostat Room Sensor from HomeKit.
- [homebridge-honeywell-home-roomsensor-thermostat](https://github.com/donavanbecker/homebridge-honeywell-home-roomsensor-thermostat)
  - The Homebridge Honeywell Home Room Sensor Thermostat plugin allows you to access your T9 Honeywell Home thermostat room sensor as a thermostat from HomeKit. This will allow you to set that room as priority in your house.
- [homebridge-honeywell-home-leaksensor](https://github.com/donavanbecker/homebridge-honeywell-leak)
  - The Homebridge Honeywell Home Leak Sensor plugin allowas you to access your Honeywell Leak Detector / Residio Droplet from HomeKit.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This project uses [Semantic Versioning](https://semver.org/)

## [Version 3.1.0](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v3.0.0...v3.1.0) (2020-11-18)

### Changes

- Fix for `Cannot set property 'ttl' of undefined`.
- Updated other dependencies.

## [Version 3.0.0](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v2.1.0...v3.0.0) (2020-09-17)

### Major Changes

- Completely reworked the way that Room Sensors discovered.

## [Version 2.1.0](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v2.0.1...v2.1.0) (2020-08-29)

### Changes

- If device is no longer Alive or Active it will automatically be removed.

## [Version 2.0.1](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v2.0.0...v2.0.1) (2020-08-24)

### Changes

- Small code changes.
- Update dependencies.

## [Version 2.0.0](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v1.2.0...v2.0.0) (2020-08-19)

### Major Changes

- Added Battery Service which allows the option to hide temperature Sensor(s).

### Changes

- Added the Option to Hide Temperature Sensor(s).
- Removed StatusActive Service.

## [Version 1.2.0](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v1.1.2...v1.2.0) (2020-08-18)

### Changes

- Added the Option to Hide Occupancy Sensor(s), Motion Sensor(s), and Humidity Sensor(s).

## [Version 1.1.2](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v1.1.1...v1.1.2) (2020-08-11)

### Changes

- Updated `package.json`.

## [Version 1.1.1](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v1.1.0...v1.1.1) (2020-08-11)

### Changes

- Updated `package.json`.

## [Version 1.1.0](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v1.0.1...v1.1.0) (2020-07-30)

### Changes

- If refresh token expires, the new refresh token that is fetched will now be saved to `config.json`.
- Changed some logging and debug logging.
- Update github links for NPM.

## [Version 1.0.1](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/compare/v1.0.0...v1.0.1) (2020-07-29)

### Changes

- Changed loggging on refresh and access token to warn logs, so they can be viewed in yellow.

## [Version 1.0.0](https://github.com/donavanbecker/homebridge-honeywell-home-roomSensor/releases/tag/v1.0.0) (2020-07-27)

### Major Changes

- This plugin allows you to be able to view your Honeywell T9/T10 Room Sensor inside HomeKit.
- This plugin is an add-on to the [Homebridge Honeywell Home plugin](https://github.com/donavanbecker/homebridge-honeywell-home).


src/settings.ts
/**
 * This is the name of the platform that users will use to register the plugin in the Homebridge config.json
 */
export const PLATFORM_NAME = 'HoneywellHomeRoomSensor';

/**
 * This must match the name of your plugin as defined the package.json
 */
export const PLUGIN_NAME = 'homebridge-honeywell-home-roomSensor';

/**
 * This is the main url used to access honeywell API
 */
export const AuthURL = 'https://api.honeywell.com/oauth2/token';

/**
 * This is the main url used to access honeywell API
 */
export const DeviceURL = 'https://api.honeywell.com/v2/devices';

/**
 * This is the main url used to access honeywell API
 */
export const LocationURL = 'https://api.honeywell.com/v2/locations';

/**
 * This is the url used to access UI Login to honeywell API
 */
export const UIurl = 'https://homebridge-honeywell.iot.oz.nu/user/refresh';


src/platformAccessory.ts
import { Service, PlatformAccessory } from 'homebridge';

import { HoneywellHomeRoomSensorPlatform } from './platform';
import { interval, Subject } from 'rxjs';
import { debounceTime, skipWhile, tap } from 'rxjs/operators';
import { DeviceURL } from './settings';

/**
 * Platform Accessory
 * An instance of this class is created for each accessory your platform registers
 * Each accessory may expose multiple services of different service types.
 */
export class RoomSensorPlatformAccessory {
  private service: Service;
  temperatureService: any;
  occupancyService: any;
  humidityService: any;
  motionService: any;

  CurrentTemperature: any;
  StatusLowBattery: any;
  OccupancyDetected: any;
  CurrentRelativeHumidity: any;
  MotionDetected!: any;
  sensor!: any;

  SensorUpdateInProgress!: boolean;
  doSensorUpdate!: any;
  TemperatureDisplayUnits!: number;


  constructor(
    private readonly platform: HoneywellHomeRoomSensorPlatform,
    private accessory: PlatformAccessory,
    public readonly locationId: string,
    public device: any,
    public findaccessories: any,
    public readonly group: any,
  ) {

    // default placeholders
    this.CurrentTemperature;
    this.StatusLowBattery;
    this.OccupancyDetected;
    this.CurrentRelativeHumidity;
    this.MotionDetected;

    // this is subject we use to track when we need to POST changes to the Honeywell API
    this.doSensorUpdate = new Subject();
    this.SensorUpdateInProgress = false;

    // set accessory information
    this.accessory.getService(this.platform.Service.AccessoryInformation)!
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Honeywell')
      .setCharacteristic(this.platform.Characteristic.Model, this.findaccessories.accessoryAttribute.type)
      .setCharacteristic(this.platform.Characteristic.SerialNumber, this.device.deviceID)
      .setCharacteristic(this.platform.Characteristic.FirmwareRevision, accessory.context.firmwareRevision);

    // get the LightBulb service if it exists, otherwise create a new LightBulb service
    // you can create multiple services for each accessory
    this.service = this.accessory.getService(this.platform.Service.BatteryService) ||
      this.accessory.addService(this.platform.Service.BatteryService),
    `${this.findaccessories.accessoryAttribute.name} Room Sensor`;

    // To avoid "Cannot add a Service with the same UUID another Service without also defining a unique 'subtype' property." error,
    // when creating multiple services of the same type, you need to use the following syntax to specify a name and subtype id:
    // this.accessory.getService('NAME') ?? this.accessory.addService(this.platform.Service.Lightbulb, 'NAME', 'USER_DEFINED_SUBTYPE');

    // set the service name, this is what is displayed as the default name on the Home app
    // in this example we are using the name we stored in the `accessory.context` in the `discoverDevices` method.
    this.service.setCharacteristic(this.platform.Characteristic.Name,
      `${this.findaccessories.accessoryAttribute.name} Room Sensor`);

    // each service must implement at-minimum the "required characteristics" for the given service type
    // see https://developers.homebridge.io/#/service/

    // Do initial device parse
    this.parseStatus();
      
    // Set Charging State
    this.service.setCharacteristic(this.platform.Characteristic.ChargingState, 2);

    // Set Low Battery
    this.service.getCharacteristic(this.platform.Characteristic.StatusLowBattery)
      .on('get', this.handeStatusLowBatteryGet.bind(this));

    // Temperature Sensor  
    this.temperatureService = accessory.getService(this.platform.Service.TemperatureSensor);  
    if (!this.temperatureService && !this.platform.config.options.hide_temperature) {
      this.temperatureService = accessory.addService(this.platform.Service.TemperatureSensor,
        `${this.findaccessories.accessoryAttribute.name} Occupancy Sensor`);

      // Set Temperature Sensor  
      this.temperatureService
        .getCharacteristic(this.platform.Characteristic.CurrentTemperature)
        .on('get', this.handleCurrentTemperatureGet.bind(this));
    } else if (this.temperatureService && this.platform.config.options.hide_temperature) {
      accessory.removeService(this.temperatureService);
    }
      
    // Occupancy Sensor  
    this.occupancyService = accessory.getService(this.platform.Service.OccupancySensor);  
    if (!this.occupancyService && !this.platform.config.options.hide_occupancy) {
      this.occupancyService = accessory.addService(this.platform.Service.OccupancySensor,
        `${this.findaccessories.accessoryAttribute.name} Occupancy Sensor`);

      // Set Occupancy Sensor  
      this.occupancyService
        .getCharacteristic(this.platform.Characteristic.OccupancyDetected)
        .on('get', this.handleOccupancyDetectedGet.bind(this));
    } else if (this.occupancyService && this.platform.config.options.hide_occupancy) {
      accessory.removeService(this.occupancyService);
    }

    // Humidity Sensor
    this.humidityService = accessory.getService(this.platform.Service.HumiditySensor);
    if (!this.humidityService && !this.platform.config.options.hide_humidity) {
      this.humidityService = accessory.addService(this.platform.Service.HumiditySensor,
        `${this.findaccessories.accessoryAttribute.name} Humidity Sensor`);

      // Set Humidity Sensor Current Relative Humidity
      this.humidityService
        .getCharacteristic(this.platform.Characteristic.CurrentRelativeHumidity)
        .on('get', this.handleCurrentRelativeHumidityGet.bind(this));
    } else if (this.humidityService && this.platform.config.options.hide_humidity) {
      accessory.removeService(this.humidityService);
    }
    
    // Motion Sensor
    this.motionService = accessory.getService(this.platform.Service.MotionSensor);
    if (!this.motionService && !this.platform.config.options.hide_motion) {
      this.motionService = accessory.addService(this.platform.Service.MotionSensor,
        `${this.findaccessories.accessoryAttribute.name} Motion Sensor`);

      // Set Motion Sensor Detected
      this.motionService
        .getCharacteristic(this.platform.Characteristic.MotionDetected)
        .on('get', this.handleMotionDetectedGet.bind(this));
    } else if (this.motionService && this.platform.config.options.hide_motion) {
      accessory.removeService(this.motionService);
    }

    // Retrieve initial values and updateHomekit
    this.refreshStatus();

    // Start an update interval
    interval(this.platform.config.options.ttl * 1000).pipe(skipWhile(() => this.SensorUpdateInProgress)).subscribe(() => {
      this.refreshStatus();
    });

    // Watch for thermostat change events
    // We put in a debounce of 100ms so we don't make duplicate calls
    this.doSensorUpdate.pipe(tap(() => {
      this.SensorUpdateInProgress = true;
    }), debounceTime(100)).subscribe(async () => {
      this.SensorUpdateInProgress = false;
    });

  }

  /**
   * Parse the device status from the honeywell api
   */
  parseStatus() {
    // Set Room Sensor State
    if (this.findaccessories.accessoryValue.batteryStatus === 'Ok') {
      this.StatusLowBattery = 0;
    } else if (this.findaccessories.accessoryValue.batteryStatus !== 'Ok') {
      this.StatusLowBattery = 1;
    }

    // Set Temperature Sensor State
    if (!this.platform.config.options.hide_temperature) {
      this.CurrentTemperature = this.toCelsius(this.findaccessories.accessoryValue.indoorTemperature);
    }

    // Set Occupancy Sensor State
    if (!this.platform.config.options.hide_occupancy) {
      if (this.findaccessories.accessoryValue.occupancyDet === true) {
        this.OccupancyDetected = 1;
      } else if (this.findaccessories.accessoryValue.occupancyDet === false) {
        this.OccupancyDetected = 0;
      }
    }

    // Set Humidity Sensor State
    if (!this.platform.config.options.hide_humidity) {
      this.CurrentRelativeHumidity = this.findaccessories.accessoryValue.indoorHumidity;
    }

    // Set Motion Sensor State
    if (!this.platform.config.options.hide_motion) {
      this.MotionDetected = this.findaccessories.accessoryValue.motionDet;
      if (this.findaccessories.accessoryValue.motionDet === false) {
        this.MotionDetected = true;
      } else if (this.findaccessories.accessoryValue.motionDet === true) {
        this.MotionDetected = false;
      }
    }
  }

  /**
   * Asks the Honeywell Home API for the latest device information
   */
  async refreshStatus() {
    try {
      const sensor = (await this.platform.axios.get(`${DeviceURL}/thermostats/${this.device.deviceID}/group/${this.group.id}/rooms`, {
        params: {
          locationId: this.locationId,
        },
      })).data;
      this.platform.log.debug(sensor);
      this.sensor = sensor;
      this.findaccessories;
      this.platform.log.debug(this.findaccessories);
      this.platform.log.debug(JSON.stringify(this.findaccessories.accessoryValue));
      this.parseStatus();
      this.updateHomeKitCharacteristics();
    } catch (e) {
      this.platform.log.error(`Failed to update status of ${this.device.name}`, e.message);
    }
  }

  /**
   * Updates the status for each of the HomeKit Characteristics
   */
  updateHomeKitCharacteristics() {
    this.service.updateCharacteristic(this.platform.Characteristic.StatusLowBattery, this.StatusLowBattery);
    if (!this.platform.config.options.hide_temperature){
      this.temperatureService.updateCharacteristic(this.platform.Characteristic.CurrentTemperature, this.CurrentTemperature);
    }
    if (!this.platform.config.options.hide_occupancy) {
      this.occupancyService.updateCharacteristic(this.platform.Characteristic.OccupancyDetected, this.OccupancyDetected);
    }
    if (!this.platform.config.options.hide_humidity) {
      this.humidityService.updateCharacteristic(this.platform.Characteristic.CurrentRelativeHumidity, this.CurrentRelativeHumidity);
    }
    if (!this.platform.config.options.hide_motion) {
      this.motionService.updateCharacteristic(this.platform.Characteristic.MotionDetected, this.MotionDetected);
    }
  }

  /**
   * Handle requests to get the current value of the "Tempeture Sensor" characteristics
   */
  handeStatusLowBatteryGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Battery Status: ${this.StatusLowBattery}`);

    // set this to a valid value for StatusLowBattery
    const currentValue = this.StatusLowBattery;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  handleCurrentTemperatureGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Current Temperature: ${this.CurrentTemperature}`);

    // set this to a valid value for CurrentTemperature
    const currentValue = this.CurrentTemperature;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  /**
   * Handle requests to get the current value of the "Occupancy Sensor" characteristics
   */
  handleOccupancyDetectedGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Occupancy: ${this.OccupancyDetected}`);

    // set this to a valid value for OccupancyDetected
    const currentValue = this.OccupancyDetected;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  /**
   * Handle requests to get the current value of the "Humidity Sensor" characteristics
   */
  handleCurrentRelativeHumidityGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Current Relative Humidity: ${this.CurrentRelativeHumidity}`);

    // set this to a valid value for CurrentRelativeHumidity
    const currentValue = this.CurrentRelativeHumidity;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  /**
   * Handle requests to get the current value of the "Motion Sensor" characteristics
   */
  handleMotionDetectedGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Motion: ${this.MotionDetected}`);

    // set this to a valid value for Motion Detected
    const currentValue = this.MotionDetected;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  /**
   * Converts the value to celsius if the temperature units are in Fahrenheit
   */
  toCelsius(value: number) {
    if (this.TemperatureDisplayUnits === this.platform.Characteristic.TemperatureDisplayUnits.CELSIUS) {
      return value;
    }

    // celsius should be to the nearest 0.5 degree
    return Math.round(((5 / 9) * (value - 32)) * 2) / 2;
  }

  /**
   * Converts the value to fahrenheit if the temperature units are in Fahrenheit
   */
  toFahrenheit(value: number) {
    if (this.TemperatureDisplayUnits === this.platform.Characteristic.TemperatureDisplayUnits.CELSIUS) {
      return value;
    }

    return Math.round((value * 9 / 5) + 32);
  }

}

src/RoomSensors/roomSensors.ts
import { Service, PlatformAccessory } from 'homebridge';
import { HoneywellHomeRoomSensorPlatform } from '../platform';
import { interval, Subject } from 'rxjs';
import { debounceTime, skipWhile, tap } from 'rxjs/operators';
import { location, sensoraccessory, T9Thermostat, T9groups } from '../configTypes';

/**
 * Platform Accessory
 * An instance of this class is created for each accessory your platform registers
 * Each accessory may expose multiple services of different service types.
 */
export class RoomSensors {
  private service: Service;
  temperatureService?: any;
  occupancyService?: any;
  humidityService?: any;
  motionService?: Service;

  CurrentTemperature!: number;
  StatusLowBattery!: number;
  OccupancyDetected!: number;
  CurrentRelativeHumidity!: number;
  MotionDetected!: any;

  SensorUpdateInProgress!: boolean;
  doSensorUpdate!: any;
  TemperatureDisplayUnits!: number;

  constructor(
    private readonly platform: HoneywellHomeRoomSensorPlatform,
    private accessory: PlatformAccessory,
    public readonly locationId: location['locationID'],
    public device: T9Thermostat,
    public sensoraccessory: sensoraccessory,
    public readonly group: T9groups,
  ) {
    // default placeholders
    this.CurrentTemperature;
    this.StatusLowBattery;
    this.OccupancyDetected;
    this.CurrentRelativeHumidity;
    this.MotionDetected;

    // this is subject we use to track when we need to POST changes to the Honeywell API
    this.doSensorUpdate = new Subject();
    this.SensorUpdateInProgress = false;

    // set accessory information
    this.accessory
      .getService(this.platform.Service.AccessoryInformation)!
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Honeywell')
      .setCharacteristic(this.platform.Characteristic.Model, this.device.deviceModel)
      .setCharacteristic(this.platform.Characteristic.SerialNumber, this.device.deviceID)
      .setCharacteristic(
        this.platform.Characteristic.FirmwareRevision,
        this.sensoraccessory.accessoryAttribute.softwareRevision,
      );

    // get the LightBulb service if it exists, otherwise create a new LightBulb service
    // you can create multiple services for each accessory
    (this.service =
      this.accessory.getService(this.platform.Service.BatteryService) ||
      this.accessory.addService(this.platform.Service.BatteryService)),
    `${this.sensoraccessory.accessoryAttribute.name} Room Sensor`;

    // To avoid "Cannot add a Service with the same UUID another Service without also defining a unique 'subtype' property." error,
    // when creating multiple services of the same type, you need to use the following syntax to specify a name and subtype id:
    // this.accessory.getService('NAME') ?? this.accessory.addService(this.platform.Service.Lightbulb, 'NAME', 'USER_DEFINED_SUBTYPE');

    // set the service name, this is what is displayed as the default name on the Home app
    // in this example we are using the name we stored in the `accessory.context` in the `discoverDevices` method.
    this.service.setCharacteristic(
      this.platform.Characteristic.Name,
      `${this.sensoraccessory.accessoryAttribute.name} Room Sensor`,
    );

    // each service must implement at-minimum the "required characteristics" for the given service type
    // see https://developers.homebridge.io/#/service/

    // Do initial device parse
    this.parseStatus();

    // Set Charging State
    this.service.setCharacteristic(this.platform.Characteristic.ChargingState, 2);

    // Set Low Battery
    this.service
      .getCharacteristic(this.platform.Characteristic.StatusLowBattery)
      .on('get', this.handeStatusLowBatteryGet.bind(this));

    // Temperature Sensor Service
    this.temperatureService = accessory.getService(this.platform.Service.TemperatureSensor);
    if (!this.temperatureService && !this.platform.config.options?.roomsensor?.hide_temperature) {
      this.temperatureService = accessory.addService(
        this.platform.Service.TemperatureSensor,
        `${this.sensoraccessory.accessoryAttribute.name} Temperature Sensor`,
      );

      // Set Temperature Sensor - Current Temperature
      this.temperatureService
        .getCharacteristic(this.platform.Characteristic.CurrentTemperature)
        .on('get', this.handleCurrentTemperatureGet.bind(this));
    } else if (this.temperatureService && this.platform.config.options?.roomsensor?.hide_temperature) {
      accessory.removeService(this.temperatureService);
    }

    // Occupancy Sensor Service
    this.occupancyService = accessory.getService(this.platform.Service.OccupancySensor);
    if (!this.occupancyService && !this.platform.config.options?.roomsensor?.hide_occupancy) {
      this.occupancyService = accessory.addService(
        this.platform.Service.OccupancySensor,
        `${this.sensoraccessory.accessoryAttribute.name} Occupancy Sensor`,
      );

      // Set Occupancy Sensor
      this.occupancyService
        .getCharacteristic(this.platform.Characteristic.OccupancyDetected)
        .on('get', this.handleOccupancyDetectedGet.bind(this));
    } else if (this.occupancyService && this.platform.config.options?.roomsensor?.hide_occupancy) {
      accessory.removeService(this.occupancyService);
    }

    // Humidity Sensor Service
    this.humidityService = accessory.getService(this.platform.Service.HumiditySensor);
    if (!this.humidityService && !this.platform.config.options?.roomsensor?.hide_humidity) {
      this.humidityService = accessory.addService(
        this.platform.Service.HumiditySensor,
        `${this.sensoraccessory.accessoryAttribute.name} Humidity Sensor`,
      );

      // Set Humidity Sensor Current Relative Humidity
      this.humidityService
        .getCharacteristic(this.platform.Characteristic.CurrentRelativeHumidity)
        .on('get', this.handleCurrentRelativeHumidityGet.bind(this));
    } else if (this.humidityService && this.platform.config.options?.roomsensor?.hide_humidity) {
      accessory.removeService(this.humidityService);
    }

    // Motion Sensor Service
    this.motionService = accessory.getService(this.platform.Service.MotionSensor);
    if (!this.motionService && !this.platform.config.options?.roomsensor?.hide_motion) {
      this.motionService = accessory.addService(
        this.platform.Service.MotionSensor,
        `${this.sensoraccessory.accessoryAttribute.name} Motion Sensor`,
      );

      // Set Motion Sensor Detected
      this.motionService
        ?.getCharacteristic(this.platform.Characteristic.MotionDetected)
        .on('get', this.handleMotionDetectedGet.bind(this));
    } else if (this.motionService && this.platform.config.options?.roomsensor?.hide_motion) {
      accessory.removeService(this.motionService);
    }

    // Retrieve initial values and updateHomekit
    this.refreshStatus();

    // Start an update interval
    interval(this.platform.config.options!.ttl! * 1000)
      .pipe(skipWhile(() => this.SensorUpdateInProgress))
      .subscribe(() => {
        this.refreshStatus();
      });

    // Watch for thermostat change events
    // We put in a debounce of 100ms so we don't make duplicate calls
    this.doSensorUpdate
      .pipe(
        tap(() => {
          this.SensorUpdateInProgress = true;
        }),
        debounceTime(100),
      )
      .subscribe(async () => {
        this.SensorUpdateInProgress = false;
      });
  }

  /**
   * Parse the device status from the honeywell api
   */
  parseStatus() {
    // Set Room Sensor State
    if (this.sensoraccessory.accessoryValue.batteryStatus.startsWith('Ok')) {
      this.StatusLowBattery = 0;
    } else {
      this.StatusLowBattery = 1;
    }

    // Set Temperature Sensor State
    if (!this.platform.config.options?.roomsensor?.hide_temperature) {
      this.CurrentTemperature = this.toCelsius(this.sensoraccessory.accessoryValue.indoorTemperature);
    }

    // Set Occupancy Sensor State
    if (!this.platform.config.options?.roomsensor?.hide_occupancy) {
      if (this.sensoraccessory.accessoryValue.occupancyDet) {
        this.OccupancyDetected = 1;
      } else {
        this.OccupancyDetected = 0;
      }
    }

    // Set Humidity Sensor State
    if (!this.platform.config.options?.roomsensor?.hide_humidity) {
      this.CurrentRelativeHumidity = this.sensoraccessory.accessoryValue.indoorHumidity;
    }

    // Set Motion Sensor State
    if (!this.platform.config.options?.roomsensor?.hide_motion) {
      this.MotionDetected = this.sensoraccessory.accessoryValue.motionDet;
    }
  }

  /**
   * Asks the Honeywell Home API for the latest device information
   */
  async refreshStatus() {
    try {
      if (this.device.deviceID.startsWith('LCC')) {
        if (this.device.deviceModel.startsWith('T9')) {
          if (this.device.groups) {
            const groups = this.device.groups;
            for (const group of groups) {
              const roomsensors = await this.platform.Sensors(this.device, group, this.locationId);
              if (roomsensors.rooms) {
                const rooms = roomsensors.rooms;
                this.platform.log.debug(JSON.stringify(roomsensors));
                for (const accessories of rooms) {
                  if (accessories) {
                    this.platform.log.debug(JSON.stringify(accessories));
                    for (const accessory of accessories.accessories) {
                      if (accessory.accessoryAttribute) {
                        if (accessory.accessoryAttribute.type) {
                          if (accessory.accessoryAttribute.type.startsWith('IndoorAirSensor')) {
                            this.sensoraccessory = accessory;
                            this.platform.log.debug(JSON.stringify(this.sensoraccessory));
                            this.platform.log.debug(JSON.stringify(this.sensoraccessory));
                            this.platform.log.debug(
                              JSON.stringify(this.sensoraccessory.accessoryAttribute.softwareRevision),
                            );
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
      this.parseStatus();
      this.updateHomeKitCharacteristics();
    } catch (e) {
      this.platform.log.error(
        `Failed to update status of ${this.sensoraccessory.accessoryAttribute.name} ${this.sensoraccessory.accessoryAttribute.type}`,
        JSON.stringify(e.message),
        this.platform.log.debug(JSON.stringify(e)),
      );
    }
  }

  /**
   * Updates the status for each of the HomeKit Characteristics
   */
  updateHomeKitCharacteristics() {
    this.service.updateCharacteristic(this.platform.Characteristic.StatusLowBattery, this.StatusLowBattery);
    if (!this.platform.config.options?.roomsensor?.hide_temperature) {
      this.temperatureService?.updateCharacteristic(
        this.platform.Characteristic.CurrentTemperature,
        this.CurrentTemperature,
      );
    }
    if (!this.platform.config.options?.roomsensor?.hide_occupancy) {
      this.occupancyService?.updateCharacteristic(
        this.platform.Characteristic.OccupancyDetected,
        this.OccupancyDetected,
      );
    }
    if (!this.platform.config.options?.roomsensor?.hide_humidity) {
      this.humidityService?.updateCharacteristic(
        this.platform.Characteristic.CurrentRelativeHumidity,
        this.CurrentRelativeHumidity,
      );
    }
    if (!this.platform.config.options?.roomsensor?.hide_motion) {
      this.motionService?.updateCharacteristic(this.platform.Characteristic.MotionDetected, this.MotionDetected);
    }
  }

  /**
   * Handle requests to get the current value of the "Tempeture Sensor" characteristics
   */
  handeStatusLowBatteryGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Battery Status: ${this.StatusLowBattery}`);

    // set this to a valid value for StatusLowBattery
    const currentValue = this.StatusLowBattery;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  handleCurrentTemperatureGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Current Temperature: ${this.CurrentTemperature}`);

    // set this to a valid value for CurrentTemperature
    const currentValue = this.CurrentTemperature;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  /**
   * Handle requests to get the current value of the "Occupancy Sensor" characteristics
   */
  handleOccupancyDetectedGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Occupancy: ${this.OccupancyDetected}`);

    // set this to a valid value for OccupancyDetected
    const currentValue = this.OccupancyDetected;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  /**
   * Handle requests to get the current value of the "Humidity Sensor" characteristics
   */
  handleCurrentRelativeHumidityGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Current Relative Humidity: ${this.CurrentRelativeHumidity}`);

    // set this to a valid value for CurrentRelativeHumidity
    const currentValue = this.CurrentRelativeHumidity;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  /**
   * Handle requests to get the current value of the "Motion Sensor" characteristics
   */
  handleMotionDetectedGet(callback: (arg0: null, arg1: any) => void) {
    this.platform.log.debug(`Update Motion: ${this.MotionDetected}`);

    // set this to a valid value for Motion Detected
    const currentValue = this.MotionDetected;

    this.doSensorUpdate.next();
    callback(null, currentValue);
  }

  /**
   * Converts the value to celsius if the temperature units are in Fahrenheit
   */
  toCelsius(value: number) {
    if (this.TemperatureDisplayUnits === this.platform.Characteristic.TemperatureDisplayUnits.CELSIUS) {
      return value;
    }

    // celsius should be to the nearest 0.5 degree
    return Math.round((5 / 9) * (value - 32) * 2) / 2;
  }

  /**
   * Converts the value to fahrenheit if the temperature units are in Fahrenheit
   */
  toFahrenheit(value: number) {
    if (this.TemperatureDisplayUnits === this.platform.Characteristic.TemperatureDisplayUnits.CELSIUS) {
      return value;
    }

    return Math.round((value * 9) / 5 + 32);
  }
}
